### ISSUE TITLE
Handling concurrent/parallel writes to object store.

### ISSUE THREAD

**Issue by brendan-m-murphy** [openghg/openghg#748]
### What is your issue?

Currently, metadata is rewritten each time data is added to the object store. This means that two files of the same data type cannot be added by two people at the same time without possibly corrupting the metadata.

**Comment by brendan-m-murphy** [comment]
Running `standardise_footprints` in parallel could cause problems because

1. metadata is overwritten
2. if there are overlapping time periods, then the new and old data need to be combined
3. even if there are no overlaps, multiple calls to `add_timed_data` (via `standardise_footprints`) could try to update the data at the same time

We could implement a map-reduce scheme to split jobs out by file (assuming date ranges don't overlap between the files being added), then process each file, including combining with existing data if there are overlapping date ranges (this can safely be done in parallel since there are no writes yet and the files we're adding don't overlap), then "reduce" the jobs by updating the datasource as the map stages complete, and also updating the metadata as part of the "reduce" stage.

To do this in parallel, we'd have to send each map job to a different batch job, which would mean the "reduce" stage is done on the login node, unless maybe the "map" outputs are stored in temporary files, which are then "reduced" in another batch job.

This still doesn't fix the issue of multiple people updating the same data type at the same time. SQL DB interfaces for Python manage to do this using "connections", which allow you to make changes locally, then tries to add them to database once the connection is closed. Multiple "connections" are managed in some way to prevent edits from happening simultaneously, which means that some changes might get rejected if your changes would conflict with changes made since you opened your connection.

"NoSQL" has something called eventual consistency for dealing with multiple people writing data at the same time, but I don't know how that works. (I'm sure there are many different schemes for resolving conflicts.)

**Comment by rt17603** [comment]
This actually seems like a duplicate of #660 but I'll close #660 instead of this issue as this has more discussion.
